finality-aleph/src/new_network/connection_manager/connections.rs

```diff
@@ -0,0 +1,85 @@
+pub struct Connections {
+    associated_sessions: HashMap<PeerId, HashSet<SessionId>>,
+    peers_by_session: HashMap<SessionId, HashSet<PeerId>>,
+}
+
+impl Connections {
+    /// Creates a new object, initially without any connections.
+    pub fn new() -> Self {
+        Connections {
+            associated_sessions: HashMap::new(),
+            peers_by_session: HashMap::new(),
+        }
+    }
+
+    /// Mark the specified peers as ones we should be connected to for the given session.
+    pub fn add_peers(&mut self, session_id: SessionId, peers: HashSet<PeerId>) {
+        for peer in peers {
+            self.associated_sessions
+                .entry(peer)
+                .or_insert_with(HashSet::new)
+                .insert(session_id);
+            self.peers_by_session
+                .entry(session_id)
+                .or_insert_with(HashSet::new)
+                .insert(peer);
+        }
+    }
+
+    /// Assume we no longer need to be connected to peers from the given session.
+    /// Returns the peers we no longer have any reason to be connected to.
+    pub fn remove_session(&mut self, session_id: SessionId) -> HashSet<PeerId> {
+        let mut result = HashSet::new();
+        if let Some(peers) = self.peers_by_session.remove(&session_id) {
+            for peer in peers {
+                if let Some(mut sessions) = self.associated_sessions.remove(&peer) {
+                    sessions.remove(&session_id);
+                    if !sessions.is_empty() {
+                        self.associated_sessions.insert(peer, sessions);
+                    } else {
+                        result.insert(peer);
+                    }
+                }
+            }
+        }
+        result
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Connections;
+    use crate::{new_network::PeerId, SessionId};
+    use sc_network::PeerId as ScPeerId;
+    use std::collections::HashSet;
+
+    fn random_peer_ids(num: usize) -> HashSet<PeerId> {
+        (0..num).map(|_| PeerId(ScPeerId::random())).collect()
+    }
+
+    #[test]
+    fn removes_peer_after_single_session() {
+        let session_id = SessionId(43);
+        let peer_ids = random_peer_ids(1);
+        let mut connections = Connections::new();
+        connections.add_peers(session_id, peer_ids.clone());
+        let to_remove = connections.remove_session(session_id);
+        assert_eq!(to_remove, peer_ids);
+    }
+
+    #[test]
+    fn does_not_remove_peer_if_still_in_session() {
+        let session_id = SessionId(43);
+        let other_session_id = SessionId(2137);
+        let peer_ids = random_peer_ids(1);
+        let mut connections = Connections::new();
+        connections.add_peers(session_id, peer_ids.clone());
+        connections.add_peers(other_session_id, peer_ids);
+        let to_remove = connections.remove_session(session_id);
+        assert!(to_remove.is_empty());
+    }
```

Comment:
Consider adding few more simple tests, e.g. similar to `does_not_remove_peer_if_still_in_session` but check if it actually removes peers after you call `.remove_session(other_session_id_`; with some other combinations (how they intersect) between two different sessions, etc.

Reply:
Added only a single test, not to overdo this. ;p

If you have more concrete suggestions let me know.